src/java.base/share/classes/jdk/internal/foreign/MemorySessionImpl.java

```diff
@@ -124,33 +159,95 @@ void addInternal(ResourceList.ResourceCleanup resource) {
         // called `ResourceList::cleanup` to run all the cleanup actions. If not, we can still add this resource
         // to the list (and, in case of an add vs. close race, it might happen that the cleanup action will be
         // called immediately after).
-        resourceList.add(resource);
+        if (resourceList != null) {
+            resourceList.add(resource);
+        }
     }
 
-    protected MemorySessionImpl(Thread owner, ResourceList resourceList) {
+    private MemorySessionImpl(Thread owner, Object ref, ResourceList resourceList) {
         this.owner = owner;
+        this.ref = ref;
         this.resourceList = resourceList;
     }
 
     public static MemorySessionImpl createConfined(Thread thread) {
-        return new ConfinedSession(thread);
+        return new MemorySessionImpl(thread, null, new ConfinedResourceList());
     }
 
     public static MemorySessionImpl createShared() {
-        return new SharedSession();
+        return new MemorySessionImpl(null, null, new SharedResourceList());
     }
 
     public static MemorySessionImpl createImplicit(Cleaner cleaner) {
-        return new ImplicitSession(cleaner);
+        var res = new MemorySessionImpl(null, null, new SharedResourceList());
+        res.state = NONCLOSEABLE;
+        VarHandle.releaseFence();
+        cleaner.register(res, res.resourceList);
+        return res;
     }
 
     public static MemorySessionImpl createHeap(Object ref) {
-        return new HeapSession(ref);
+        Objects.requireNonNull(ref);
+        var res = new MemorySessionImpl(null, ref, null);
+        res.state = NONCLOSEABLE;
+        VarHandle.releaseFence();
+        return res;
     }
 
-    public abstract void release0();
+    public void release0() {
+        if (owner == Thread.currentThread()) {
+            acquireCount--;
+            return;
+        } else if (owner != null) {
+            // It is possible to end up here in two cases: this session was kept
+            // alive by some other confined session which is implicitly released
+            // (in which case the release call comes from the cleaner thread).
+            // Or, this session might be kept alive by a shared session, which
+            // means the release call can come from any thread.
+            int a = (int)ASYNC_RELEASE_COUNT.getAndAdd(this, 1);
+            return;
+        }
 
-    public abstract void acquire0();
+        VarHandle.acquireFence();
+        if (state == NONCLOSEABLE) {
+            Reference.reachabilityFence(this);
+            return;
+        }
+
+        int a = (int)ACQUIRE_COUNT.getAndAdd(this, -1);
+    }
+
+    public void acquire0() {
+        if (owner == Thread.currentThread()) {
```

Comment:
How much does it help here to have a single implementation? Effectively here you trade off a branch (and more unreadable code) with bimporphic inlining? (e.g. if there's just shared and confined implementations, I think it should still inline well, of course except in cases where the profile is polluted).

I suppose my more general question is: is there some evidence that the thing being optimized here is problematic?

Reply:
`acquire0` has multiple implementations so in theory there can be a megamorphism here. In addition, in a separate `ConfinedSession`, we still need to check `owner == Thread.currentThread()` anyway, so this may save 1 branch at worst and a virtual invocation at best.

Reply:
Also, I have extracted code for each type of session into a separate method to improve readability.

Reply:
I was about to suggest something similar